File: haskell/src/Shapes/Base/Models/Circle.hs
module Shapes.Base.Models.Circle (
    Circle (..),
    newCircle,
    getColor,
    setColor,
) where

import Data.Char
import Data.List (intercalate)
import Shapes.Base.Commons
import Text.Printf

data Circle where
    Circle :: {radius :: !Double, color :: !String} -> Circle
    deriving (Eq, Show)

instance Dimensional Circle where
    area (Circle radius _) = pi * radius * radius

    perimeter (Circle radius _) = 2 * pi * radius

instance Displayable Circle

instance BaseShape Circle where
    dimensions (Circle radius _) =
        [ ("radius", max radius 1),
          ("diameter", 2 * max radius 1),
          ("curvature", 1 / max radius 1)
        ]

    printDescription circle = do
        putStrLn ("Circle radius: " ++ show (radius circle))
        putStrLn ("Circle dimensions: " ++ formatDimensions (dimensions circle))
        putStrLn ("Circle color: " ++ color circle)
        putStrLn ("Circle area: " ++ show (area circle))
        putStrLn ("Circle perimeter: " ++ show (perimeter circle))
        putStrLn ("Circle is a displayable shape: " ++ show (shouldDisplay circle))
        putStrLn ("Circle is a shape: " ++ show (validateShape circle))

newCircle :: Double -> Circle
newCircle radius = Circle (max radius 1) "None"

getColor :: Circle -> String
getColor (Circle _ color) = color

setColor :: Circle -> String -> Circle
setColor (Circle radius _) = Circle radius

formatDimensions :: [(String, Double)] -> String
formatDimensions dimension = intercalate " | " parts
  where
    parts = [map toUpper key ++ ": " ++ printf "%.3f" value | (key, value) <- dimension]

File: csharp/Src/Shapes/Base/Models/Circle.cs
using Shapes.Base.Commons;

namespace Shapes.Base.Models;

public sealed class Circle : BaseShape, IDimensional, IDisplayable {
    private readonly double _radius;
    private string _color;

    public Circle(double radius)
        : base(
            [
                new() { { "radius", Math.Max(radius, 1) } },
                new() { { "diameter", Math.Max(radius, 1) * 2 } },
                new() { { "curvature", 1 / Math.Max(radius, 1) } },
            ]
        ) {
        double dim1 = radius > 0 ? radius : 1;

        _radius = dim1;
        _color = "None";
    }

    public string Color {
        get {
            return _color;
        }

        set {
            _color = value;
        }
    }

    public override void PrintDescription() {
        Console.WriteLine($"Circle radius: {_radius}");
        Console.WriteLine($"Circle dimensions: {FormatDimensions()}");
        Console.WriteLine($"Circle color: {Color}");
        Console.WriteLine($"Circle area: {Area()}");
        Console.WriteLine($"Circle perimeter: {Perimeter()}");
        Console.WriteLine($"Circle is displayable: {((IDisplayable)this).ShouldDisplay()}");
        Console.WriteLine($"Circle is a shape: {ValidateShape()}");
    }

    public double Area() {
        return Math.PI * Math.Pow(_radius, 2);
    }

    public double Perimeter() {
        return 2 * Math.PI * _radius;
    }

    private string FormatDimensions() {
        IEnumerable<string> parts = Dimensions.Select(dim =>
            string.Join("", dim.Select(kv => $"{kv.Key}: {Math.Round(kv.Value, 3):F3}"))
        );

        return string.Join(" | ", parts);
    }
}
